### ISSUE TITLE
Proxy IP forwarding support

### ISSUE THREAD

**Issue by caelunshun** [feather-rs/feather#155]
Support for IP forwarding with BungeeCord/Velocity proxies.

For Velocity: https://github.com/PaperMC/Paper/pull/1557

BungeeCord sends the following example data in the `server_address` field in the Handshake packet:
```
localhost\0127.0.0.1\0b26a922dc0604ed5a116e0372a349f1c\0[{"name":"textures","value":"eyJ0aW1lc3RhbXAiOjE1NzAzMDkwMTYxMjAsInByb2ZpbGVJZCI6ImIyNmE5MjJkYzA2MDRlZDVhMTE2ZTAzNzJhMzQ5ZjFjIiwicHJvZmlsZU5hbWUiOiJjYWVsdW5zaHVuIiwidGV4dHVyZXMiOnsiU0tJTiI6eyJ1cmwiOiJodHRwOi8vdGV4dHVyZXMubWluZWNyYWZ0Lm5ldC90ZXh0dXJlLzNjMWE0OWMyZTlmYmYzMjM1YTQwZDI3NWFhMTlhYjBjOTFkYWViN2EyMDM5Njk0ODMzMmJiODc3ZTBlODAzYjUifX19","signature":"S1O105istJx5CrAVq2vpE6WO2qWzMCS6SWRyAanaI8zAr3m+iB7fEivEoVZm1Y+sORb+adDR29jxeYkZyuU3ADF5OQr/pCL46Tk/Kolt72/9i7TWT5uUKT3K/JFA/bHgidDo2AxhdsuKK3zNG12//S/ZiVNfkTlUSNF6u7tS0fVH35gArwzKztzvXDej4JNg+8UiLi2kmL70Mq99KhRHl4GLWzG91gGihueaIDJQNf7QjBhfM28oLmtVHXODpzHdFDgfwVgETQjCQfz4rrDlPvxfNXEi7HQT/mX+hjdx2E9mB2Lan8sZApOA2qlEV79LEMqtpy5wuVFWYPgSobblN2Fpd9lb1jAff5fmk/Kw1NmQ0CoR5B8dh5HHZsxJymwjrr6DAEO4QM/A/Lq+fSo7wjwnZqJiJkSelyi7q0RRUo3d/jbJYf/vn3zR3e90pccH5YO38W338YPMTncLekfZWzEdtEFxgLj42PGSuTfvhASMxC0GTUHcoT/8a8Z2ShrHjrbuav9Sn4xJSHSnwO0FA7Cx/rZ7dWh4o4qezWOOD8pY+Yp1zcacGRQODIRawA83+S9WPu0NfJJVhE0UC4L/zh5SJAI1RM4+nYEdkMePbwesMUTGZPyds9KxQPyVF9vzUyvzC2wY6g/0zLZ8EIYK5idPr6UZohtHc+3n7geJLTk\u003d"}]
```

**Comment by ThijsRay** [comment]
I tried to collect some more information about the exact information that is sent in the Handshake packet, in case that IP forwarding is enabled.

### BungeeCord
The lines of code that are responsible for the altered Handshake packet can be found in BungeeCords [ServerConnector.java](https://github.com/SpigotMC/BungeeCord/blob/1b34a16c097d23bb89b5cdd5f7a6d52ff37e3798/proxy/src/main/java/net/md_5/bungee/ServerConnector.java#L99) file.

They add more information `server_address` field of the Handshake packet. It has the following format:

```rust
format!("{}\0{}\0{}\0{}", host, address, uuid, mojang_response);
```
| Variable | Definition |
|------------|------------|
| Host | The IP address of the BungeeCord instance |
| Address | The IP address of the connecting client |
| UUID | The UUID that is associated to the clients account |
| Mojang response| A JSON formatted version of the `properties` field in [Mojangs response](https://wiki.vg/Protocol_Encryption#Server) |

Example
```
192.168.1.87\0192.168.1.67\0f68a68e9b28c44a1aee749a59117567d\0[{"name":"textures","value":"<base64 string>","signature":"<base64 string>"}]
```

Whenever a client is connecting through a BungeeCord proxy, offline mode is enabled on the server. The UUIDs generated in offline mode do not match the UUIDs that are generated in online mode. To ensure that there is consistent behavior between different servers behind a BungeeCord proxy, the UUID that is generated in offline mode should be changed to the UUID that is sent by BungeeCord.

The server should probably check if the given host IP address
* Matches the IP address of the sender of the packet
* (Maybe?) Check if the host IP address is on a whitelist. This should probably be optional, as this doesn't scale very well.

----

I'll will update this later for information about IP forwarding with Velocity proxies.

**Comment by caelunshun** [comment]
Thanks for the information!

If anyone wants to implement this, the file to look at is `server/src/io/initialhandler.rs`—it contains the code which performs the protocol handshake and eventually returns the player's UUID and profile properties. Currently, it generates a random UUID when in offline mode—if IP forwarding is enabled, the Handshake packet should be parsed to obtain the valid UUID, profile, and host, as expected.

**Comment by momothereal** [comment]
Is this done by #165 ?

**Comment by caelunshun** [comment]
> Is this done by #165 ?

BungeeCord support was added, but Velocity isn't implemented yet.

**Comment by Stupremee** [comment]
I would like to implement the velocity part.

**Comment by Defman** [comment]
Should such functionality not be implemented through plugins?

**Comment by caelunshun** [comment]
> Should such functionality not be implemented through plugins?

It certainly could be, although most server implementations have built-in support for proxy IP forwarding. What should and should not be delegated to plugins is an ongoing discussion, though.

**Comment by caelunshun** [comment]
Implemented in #360.
